common/parallel/TreeCrawler.go

```diff
@@ -168,21 +309,39 @@ func (c *crawler) processOneDirectory(ctx context.Context, workerIndex int) (boo
 	addDir := func(d Directory) {
 		foundDirectories = append(foundDirectories, d)
 	}
+
 	addOutput := func(de DirectoryEntry, er error) {
 		select {
 		case c.output <- CrawlResult{item: de, err: er}:
 		case <-ctx.Done(): // don't block on full channel if cancelled
 		}
 	}
+
 	bodyErr := c.workerBody(toExamine, addDir, addOutput) // this is the worker body supplied by our caller
 
 	// finally, update shared state (inside the lock)
 	c.cond.L.Lock()
 	defer c.cond.L.Unlock()
 
-	c.unstartedDirs = append(c.unstartedDirs, foundDirectories...) // do NOT try to wait here if unstartedDirs is getting big. May cause deadlocks, due to all workers waiting and none processing the queue
-	c.dirInProgressCount--                                         // we were doing something, and now we have finished it
-	c.cond.Broadcast()                                             // let other workers know that the state has changed
+	//
+	// Source traverser MUST add all completed directories to tqueue, for target traverser to process.
+	// Note that c.workerBody() above would have completed enumeration of 'toExamine' dir and hence this is
+	// the right time to add it to tqueue.
+	//
+	if c.isSync && c.isSource {
+		if _, ok := toExamine.(string); ok {
+			if _, ok := c.root.(string); ok {
```

Comment:
why do we need to do the c.root.(string) check everytime? we should do it once in the beginning?

Reply:
In case of azure file , root is not string. It's directory url. We can make it work with string but that's separate change.